packages/synapse-react-client/src/components/SetAccessRequirementCommonFields/SetAccessRequirementCommonFields.tsx

```diff
@@ -0,0 +1,309 @@
+    )
+    const [clientError, setClientError] = useState<string | null>(null)
+
+    const onMutationSuccess = (ar: AccessRequirement) => {
+      setClientError(null)
+      onSave(ar.id.toString(), ar.concreteType)
+    }
+
+    const onMutationError = (error: SynapseClientError) => {
+      setClientError(error.reason)
+      onError()
+    }
+
+    const { mutate: createAccessRequirement } = useCreateAccessRequirement({
+      onSuccess: newAr => onMutationSuccess(newAr),
+      onError: error => onMutationError(error),
+    })
+
+    const { mutate: updateAccessRequirement } = useUpdateAccessRequirement({
+      onSuccess: updatedAr => onMutationSuccess(updatedAr),
+      onError: error => onMutationError(error),
+    })
+
+    const {
+      data: accessRequirement,
+      isLoading: isLoadingAr,
+      error: getArError,
+    } = useGetAccessRequirements(accessRequirementId!, {
+      enabled: Boolean(accessRequirementId),
+    })
+
+    useEffect(() => {
+      if (accessRequirement) {
+        setArType(accessRequirement.concreteType)
+        setName(accessRequirement.name)
+        setSubjects(accessRequirement.subjectIds)
+      }
+    }, [accessRequirement])
+
+    const subjectsType = useMemo(() => {
+      if (subject) {
+        return subject.type
+      }
+      if (accessRequirement) {
+        const initialSubjects = accessRequirement.subjectIds
+        if (initialSubjects.length > 0) {
+          return initialSubjects[0].type
+        }
+      }
+      return null
+    }, [subject, accessRequirement])
+
+    const selectorContent = useMemo(() => {
+      if (!subjectsType) return <></>
+
+      function onUpdate(subjects: RestrictableObjectDescriptor[]) {
+        setSubjectsError(null)
+        setSubjects(subjects)
+      }
+
+      function onUpdateIdsTextbox(value: string) {
+        setSubjectsError(null)
+        setHasPendingSubjects(value.trim() !== '')
+      }
+
+      switch (subjectsType) {
+        case RestrictableObjectType.TEAM:
+          return (
+            <TeamSubjectsSelector
+              subjects={subjects}
+              onUpdate={subjects => onUpdate(subjects)}
+              onUpdateTeamIDsTextbox={value => onUpdateIdsTextbox(value)}
+            />
+          )
+        case RestrictableObjectType.ENTITY:
+          return (
+            <EntitySubjectsSelector
+              subjects={subjects}
+              onUpdate={subjects => onUpdate(subjects)}
+              onUpdateEntityIDsTextbox={value => onUpdateIdsTextbox(value)}
+            />
+          )
+        default:
+          console.error(
+            `RestrictableObjectType ${subjectsType} does not have a selector implemented.`,
+          )
+          return <></>
+      }
+    }, [subjectsType, subjects])
+
+    useImperativeHandle(
+      ref,
+      () => {
+        return {
+          save() {
+            const isStillLoading =
+              (isEditing && !accessRequirement) || !subjectsType
```

Comment:
Handle case where user clicks "Save" in the parent before this component has finished fetching the AR